core/src/main/java/io/grpc/internal/TransportSet2.java

```diff
@@ -0,0 +1,501 @@
+        if (log.isLoggable(Level.FINE)) {
+          log.log(Level.FINE, "[{0}] Terminated in shutdown()", getLogId());
+        }
+      }  // else: the callback will be run once all transports have been terminated
+      cancelReconnectTask();
+    }
+    if (savedActiveTransport != null) {
+      savedActiveTransport.shutdown();
+    }
+    if (savedPendingTransport != null) {
+      savedPendingTransport.shutdown();
+    }
+  }
+
+  // May be called under lock.
+  private void handleTermination() {
+    channelExecutor.execute(new Runnable() {
+        @Override
+        public void run() {
+          callback.onTerminated(TransportSet2.this);
+        }
+      });
+  }
+
+  private void handleTransportInUseState(
+      final ManagedClientTransport transport, final boolean inUse) {
+    channelExecutor.execute(new Runnable() {
+        @Override
+        public void run() {
+          inUseStateAggregator.updateObjectInUse(transport, inUse);
+        }
+      });
+  }
+
+  void shutdownNow(Status reason) {
+    shutdown();
+    Collection<ManagedClientTransport> transportsCopy;
+    synchronized (lock) {
+      transportsCopy = new ArrayList<ManagedClientTransport>(transports);
+    }
+    for (ManagedClientTransport transport : transportsCopy) {
+      transport.shutdownNow(reason);
+    }
+  }
+
+  @GuardedBy("lock")
+  private void cancelReconnectTask() {
+    if (reconnectTask != null) {
+      reconnectTask.cancel(false);
+      reconnectTask = null;
+    }
+  }
+
+  @Override
+  public String getLogId() {
+    return GrpcUtil.getLogId(this);
+  }
+
+  @VisibleForTesting
+  ConnectivityState getState() {
+    synchronized (lock) {
+      return state.getState();
+    }
+  }
+
+  /** Listener for real transports. */
+  private class TransportListener implements ManagedClientTransport.Listener {
+    final ManagedClientTransport transport;
+    final SocketAddress address;
+
+    TransportListener(ManagedClientTransport transport, SocketAddress address) {
+      this.transport = transport;
+      this.address = address;
+    }
+
+    @Override
+    public void transportReady() {
+      if (log.isLoggable(Level.FINE)) {
+        log.log(Level.FINE, "[{0}] {1} for {2} is ready",
+            new Object[] {getLogId(), transport.getLogId(), address});
+      }
+      ConnectivityState savedState;
+      synchronized (lock) {
+        savedState = state.getState();
+        reconnectPolicy = null;
+        nextAddressIndex = 0;
+        if (savedState == SHUTDOWN) {
+          // activeTransport should have already been set to null by shutdown(). We keep it null.
+          Preconditions.checkState(activeTransport == null,
+              "Unexpected non-null activeTransport");
+        } else if (pendingTransport == transport) {
+          gotoNonErrorState(READY);
+          activeTransport = transport;
+          pendingTransport = null;
+        }
+      }
+      if (savedState == SHUTDOWN) {
+        transport.shutdown();
```

Comment:
Question: Is it OK to call `shutdown()` more than once for a given transport? I was wondering of the following case:

1) a pending transport is created.
2) `#shutdown()` is called. It calls `pendingTransport.shutdown()` and sets `pendingTransport` to null.
3) `transportReady()` is called, and `transport.shutdown()` is called again.